networkx/algorithms/community/demon.py

```diff
@@ -0,0 +1,236 @@
+    for node in labels:
+        n_labels = labels[node]
+        for label in n_labels:
+            community_to_nodes[label].append(node)
+            
+    for c in community_to_nodes:
+        community_to_nodes[c].append(ego)
+
+    return community_to_nodes
+
+
+def _merge_communities(communities, actual_community, epsilon):
+    """
+    Builds communities using the incomplete communities detected by the label
+    propagation algorithm in the EgoMinusEgo graphs.
+
+    Parameters
+    ----------
+    communities: dict
+        Dictionary of communities
+
+    actual_community: set
+        A community
+
+    epsilon: real
+        The tolerance required in order to merge communities.
+
+    Returns
+    -------
+    communities : list
+        Merged communities
+    """
+
+    # if the community is already present return
+    if actual_community in communities:
+        return communities
+
+    else:
+        # search a community to merge with
+        for test_community in communities:
+            union = _generalized_inclusion(actual_community, test_community, epsilon)
+            # community to merge with found
+            if union:
+                communities.pop(communities.index(test_community))
+                communities.append(union)
+                break
+        # not merged: insert the original community
+        else:
+            communities.append(actual_community)
+
+    return communities
+
+
+def _generalized_inclusion(c1, c2, epsilon):
+    """
+    As explained in [1]_:
+    "Two  communities C and I are  merged  if  and  only  if  at most the
+    epsilon % of the smaller one is not included in the bigger one;  in  this
+    case, C and I are removed from the community list and their union is added
+    to the result set. The epsilon factor is introduced to vary the percentage
+    of common elements provided from each couple of communities: epsilon = 0
+    ensure that two communities are merged only if one of them is a proper
+    subset of the other, on the other hand with a value of epsilon = 1 even
+    communities that do not share a single node are merged together"
+
+    Parameters
+    ----------
+    c1 : set
+        Community 1.
+
+    c2 : set
+        Community 2.
+
+    epsilon: real
+        The tolerance required in order to merge communities
+
+    Returns
+    -------
+    union : set
+        Union of both communities if the similarity is greater than epsilon
+
+    References
+    ----------
+    .. [1] Michele Coscia, Giulio Rossetti, Fosca Giannotti, Dino Pedreschi:
+           DEMON: a local-first discovery method for overlapping communities.
+           KDD 2012:615-623.
+    """
+    intersection = c2 & c1
+    smaller_set_size = min(len(c1), len(c2))
+
+    if len(intersection) > 0 and smaller_set_size > 0:
+        inclusion_pct = len(intersection) / smaller_set_size
+
+        if inclusion_pct > epsilon:
+            union = c2 | c1
+            return union
+
+    return None
```

Comment:
How about this for a simpler implementation:

    intersection = c1 & c2
    all_nonempty = intersection and c1 and c2
    if all_nonempty and len(intersection) / min(len(c1), len(c2)) > epsilon:
            return c1 | c2
    return None